db/commitlog/commitlog.cc

```diff
@@ -708,40 +733,38 @@ class db::commitlog::segment : public enable_shared_from_this<segment>, public c
     }
     future<sseg_ptr> close() {
         _closed = true;
-        return sync().then([] (sseg_ptr s) {
-            return s->flush();
-        }).then([](sseg_ptr s) {
-            return s->terminate();
-        }).then([](sseg_ptr s) {
-            s->_segment_manager->totals.wasted_size_on_disk += (s->_size_on_disk - s->file_position());
-            return s;
-        });
+        auto s = co_await sync();
+        co_await flush();
```

Comment:
It looks like you rely on `s == shared_from_this()`.
If so, why bother to keep the `sseg_ptr` that `sync()` returns?

Why keep passing the sseg_ptr as return values?
(https://github.com/scylladb/scylla/issues/8952 shows it might not even work in all cases)

Reply:
It is pure self-keepalive. Because we run some of these routines in non-anchored places (i.e. let the continuation run without waiting). One potentially great future refactoring would perhaps be to remove that particular behaviour.